hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,194 @@
+
+// typeWalkerRemoveType is a special TypeName that informs the type walker to remove the property containing this TypeName
+// entirely.
+var typeWalkerRemoveType = MakeTypeName(LocalPackageReference{}, "TypeWalkerRemoveProperty")
+
+// TODO: it's awkward to have so much configuration on this thing (3 separate funcs that apply at different places in the walking process?)
+// TODO: but unsure if there's a better way... bring it up in code review.
+
+// TypeWalker performs a depth first traversal across the types provided, applying the visitor to each TypeDefinition.
+// MakeContext is called before each visit, and AfterVisit is called after each visit. ShouldRemoveCycle is called
+// if a cycle is detected.
+type TypeWalker struct {
+	allTypes Types
+	visitor  TypeVisitor
+
+	// MakeContext is called before any type but the root type is visited. It is given the current context and returns
+	// a new context for use in the upcoming Visit.
+	MakeContext func(it TypeName, ctx interface{}) (interface{}, error)
+	// AfterVisit is called after the type walker has applied the visitor to a TypeDefinition.
+	AfterVisit func(original TypeDefinition, updated TypeDefinition, ctx interface{}) (TypeDefinition, error)
+	// ShouldRemoveCycle is called if a cycle is detected. If true is returned the cycle will be pruned, otherwise it will be preserved as-is.
+	ShouldRemoveCycle func(def TypeDefinition, ctx interface{}) (bool, error)
+
+	state                   typeWalkerState
+	originalVisitTypeName   func(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error)
+	originalVisitObjectType func(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error)
+}
+
+type typeWalkerState struct {
+	result     Types
+	processing map[TypeName]struct{}
+}
+
+// NewTypeWalker returns a TypeWalker.
+// The provided visitor VisitTypeName function must return a TypeName and VisitObjectType must return an ObjectType or calls to
+// Walk will panic.
+func NewTypeWalker(allTypes Types, visitor TypeVisitor) *TypeWalker {
+	typeWalker := TypeWalker{
+		allTypes:                allTypes,
+		originalVisitTypeName:   visitor.VisitTypeName,
+		originalVisitObjectType: visitor.VisitObjectType,
+		AfterVisit:              IdentityAfterVisit,
+		MakeContext:             IdentityMakeContext,
+		ShouldRemoveCycle:       IdentityShouldRemoveCycle,
+	}
+
+	// visitor is a copy - modifications won't impact passed visitor
+	visitor.VisitTypeName = typeWalker.visitTypeName
+	visitor.VisitObjectType = typeWalker.visitObjectType
+
+	typeWalker.visitor = visitor
+
+	return &typeWalker
+}
+
+func (t *TypeWalker) visitTypeName(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error) {
+	updatedCtx, err := t.MakeContext(it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	visitedTypeName, err := t.originalVisitTypeName(this, it, updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+	it, ok := visitedTypeName.(TypeName)
```

Comment:
Do you want exactly a `TypeName` or should you use `AsTypeName()`?

Reply:
I want exactly a `TypeName`. As mentioned in the comment on line 46:
> // The provided visitor VisitTypeName function must return a TypeName and VisitObjectType must return an ObjectType or calls to Walk will panic.

Reply:
Fair enough.